<a href="https://colab.research.google.com/github/samsomsabu/NATURAL-LANGUAGE-PROCESSING/blob/main/2348549_SAMSON_ESE_COMPONENT_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Implement the Spam Classification using

a. RNN


b. LSTM

IMPORT THE  LIBRARIES NEEDED

In [76]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Embedding, LSTM, SimpleRNN, Dense
from tensorflow.keras.callbacks import EarlyStopping

LOAD THE DATASET

In [63]:
df = pd.read_csv("/content/spam.csv",encoding = "ISO-8859-1")

---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-39-20a8c0c3b243> in <cell line: 1>()
----> 1 df = pd.read_csv("/content/spam.csv")

9 frames
/usr/local/lib/python3.10/dist-packages/pandas/_libs/parsers.pyx in pandas._libs.parsers.raise_parser_error()

UnicodeDecodeError: 'utf-8' codec can't decode bytes in position 606-607: invalid continuation byte

having issues with loading the csv file . so used the solution found from "stacksoverflow" encoding = "ISO-8859-1"

Read the first five datapoints

In [64]:
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

It was found that there was three unnecessary columns so created a dataframe with only the v1 and v2. Renamed them into Label and Text directly

In [65]:
# Removing unnecessary columns
df = df[['v1', 'v2']]

# Renaming columns
df.columns = ['label', 'text']

# Displaying the updated DataFrame
print(df.head())


  label                                               text
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...


checking for null values

In [66]:
df.isnull().sum()

label    0
text     0
dtype: int64

adding input variable to X and taarget variable to Y

In [67]:
# Preprocessing
X = df['text']
y = pd.get_dummies(df['label'], drop_first=True)


tokenizing, converting to sequences, padding the sequences

In [68]:
tokenizer = Tokenizer()
tokenizer.fit_on_texts(X)
X_seq = tokenizer.texts_to_sequences(X)
X_pad = pad_sequences(X_seq)

 splitting the data into training and testing sets

In [69]:
X_train, X_test, y_train, y_test = train_test_split(X_pad, y, test_size=0.2, random_state=42)

MODEL BUILDING

In [70]:
model_rnn = Sequential()
model_rnn.add(Embedding(input_dim=len(tokenizer.word_index)+1, output_dim=100, input_length=X_pad.shape[1]))
model_rnn.add(SimpleRNN(units=64))
model_rnn.add(Dense(1, activation='sigmoid'))

Compiling model

In [71]:
model_rnn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

Training the model or fitting

In [72]:
model_rnn.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=64, callbacks=[EarlyStopping(monitor='val_loss', patience=2)])


Epoch 1/10
70/70 [==============================] - 20s 264ms/step - loss: 0.3775 - accuracy: 0.8708 - val_loss: 0.1588 - val_accuracy: 0.9623
Epoch 2/10
70/70 [==============================] - 18s 262ms/step - loss: 0.0697 - accuracy: 0.9827 - val_loss: 0.0704 - val_accuracy: 0.9821
Epoch 3/10
70/70 [==============================] - 16s 228ms/step - loss: 0.0226 - accuracy: 0.9946 - val_loss: 0.0684 - val_accuracy: 0.9839
Epoch 4/10
70/70 [==============================] - 13s 187ms/step - loss: 0.0078 - accuracy: 0.9991 - val_loss: 0.0699 - val_accuracy: 0.9830
Epoch 5/10
70/70 [==============================] - 13s 185ms/step - loss: 0.0038 - accuracy: 0.9993 - val_loss: 0.0693 - val_accuracy: 0.9839


The model was trained for 5 epochs

The training loss decreased significantly from 0.3775 to 0.0038, while the training accuracy steadily increased, reaching close to 100% (0.9993)

The validation loss and accuracy show good performance, with the loss  around 0.07 and the accuracy  around 98-99%.

this means there is convergence but there is also no issue of overfitting because validation metrics are somewhat similar to the training ones

building lstm model

In [73]:
model_lstm = Sequential()
model_lstm.add(Embedding(input_dim=len(tokenizer.word_index)+1, output_dim=100, input_length=X_pad.shape[1]))
model_lstm.add(LSTM(units=64))
model_lstm.add(Dense(1, activation='sigmoid'))

compiling lstm model and fitting or training it

In [74]:
model_lstm.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

model_lstm.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=10, batch_size=64, callbacks=[EarlyStopping(monitor='val_loss', patience=2)])

Epoch 1/10
70/70 [==============================] - 9s 107ms/step - loss: 0.2515 - accuracy: 0.9145 - val_loss: 0.0804 - val_accuracy: 0.9803
Epoch 2/10
70/70 [==============================] - 6s 79ms/step - loss: 0.0367 - accuracy: 0.9886 - val_loss: 0.0583 - val_accuracy: 0.9839
Epoch 3/10
70/70 [==============================] - 3s 41ms/step - loss: 0.0125 - accuracy: 0.9966 - val_loss: 0.0542 - val_accuracy: 0.9839
Epoch 4/10
70/70 [==============================] - 3s 41ms/step - loss: 0.0033 - accuracy: 0.9996 - val_loss: 0.0664 - val_accuracy: 0.9839
Epoch 5/10
70/70 [==============================] - 2s 21ms/step - loss: 0.0012 - accuracy: 0.9998 - val_loss: 0.0699 - val_accuracy: 0.9830


The model was trained for 5 epochs

The training loss decreased significantly from 0.2515 to 0.0012, while the training accuracy steadily increased to close to 100% (0.9998).

The training loss decreased from 0.2515 to 0.0012, while the training accuracy  increased to ery close to 100% (0.9998).

covergencce is there same for overfitting

EVALUATING EACH MODEL SEPARATELY

In [ ]:
_, rnn_accuracy = model_rnn.evaluate(X_test, y_test)
print(f"RNN Accuracy: {rnn_accuracy}")

_, lstm_accuracy = model_lstm.evaluate(X_test, y_test)
print(f"LSTM Accuracy: {lstm_accuracy}")


35/35 [==============================] - 1s 16ms/step - loss: 0.0693 - accuracy: 0.9839
RNN Accuracy: 0.9838564991950989
35/35 [==============================] - 0s 5ms/step - loss: 0.0699 - accuracy: 0.9830
LSTM Accuracy: 0.9829596281051636


The accuracy results for the RNN and LSTM models on the test dataset are as follows:

RNN Accuracy: 0.9839


LSTM Accuracy: 0.9830